src/addr.rs

```diff
@@ -1,117 +1,251 @@
+            pub const fn align_down_4k(self) -> Self {
+                Self::from_usize(align_down(self.0, PAGE_SIZE_4K))
+            }
+
+            /// Aligns the address upwards to 4096 (bytes).
+            #[inline]
+            pub const fn align_up_4k(self) -> Self {
+                Self::from_usize(align_up(self.0, PAGE_SIZE_4K))
+            }
+
+            /// Returns the offset of the address within a 4K-sized page.
+            #[inline]
+            pub const fn align_offset_4k(self) -> usize {
+                align_offset(self.0, PAGE_SIZE_4K)
+            }
+
+            /// Checks whether the address is 4K-aligned.
+            #[inline]
+            pub const fn is_aligned_4k(self) -> bool {
+                is_aligned(self.0, PAGE_SIZE_4K)
+            }
+        }
+
+        impl From<usize> for $name {
+            #[inline]
+            fn from(addr: usize) -> Self {
+                Self(addr)
+            }
+        }
+
+        impl From<$name> for usize {
+            #[inline]
+            fn from(addr: $name) -> usize {
+                addr.0
+            }
+        }
+
+        impl core::ops::Add<usize> for $name {
+            type Output = Self;
+            #[inline]
+            fn add(self, rhs: usize) -> Self {
+                Self(self.0 + rhs)
+            }
+        }
+
+        impl core::ops::AddAssign<usize> for $name {
+            #[inline]
+            fn add_assign(&mut self, rhs: usize) {
+                self.0 += rhs;
+            }
+        }
+
+        impl core::ops::Sub<usize> for $name {
+            type Output = Self;
+            #[inline]
+            fn sub(self, rhs: usize) -> Self {
+                Self(self.0 - rhs)
+            }
+        }
+
+        impl core::ops::SubAssign<usize> for $name {
+            #[inline]
+            fn sub_assign(&mut self, rhs: usize) {
+                self.0 -= rhs;
+            }
+        }
+
+        $crate::def_usize_addr!($($tt)*);
+    };
+    () => {};
+}
 
-    /// Aligns the address downwards to 4096 (bytes).
-    #[inline]
-    pub const fn align_down_4k(self) -> Self {
-        Self(align_down(self.0, PAGE_SIZE_4K))
-    }
+/// Creates implementations for the [`core::fmt::Debug`], [`core::fmt::LowerHex`], and
+/// [`core::fmt::UpperHex`] traits for the given address types defined by the [`def_usize_addr`].
+///
+/// For each `$name = $format;`, this macro generates the following items:
+/// - An implementation of [`core::fmt::Debug`] for the address type `$name`, which formats the
+///   address with `format_args!($format, format_args!("{:#x}", self.0))`,
+/// - An implementation of [`core::fmt::LowerHex`] for the address type `$name`, which formats the
+///   address in the same way as [`core::fmt::Debug`],
+/// - An implementation of [`core::fmt::UpperHex`] for the address type `$name`, which formats the
+///   address with `format_args!($format, format_args!("{:#X}", self.0))`.
+///
+/// # Example
+///
+/// ```
```

Comment:
Add example code to show the usage of this macro.

Reply:
Added